### ISSUE TITLE
[BUG] Possible Deadlock Isue on A2A_Overlap overlap-moe-expert-parallel-comm during early MoE training on large runs

### ISSUE THREAD

**Issue by Skylion007** [NVIDIA/Megatron-LM#1810]
Deadlock when `--overlap-moe-expert-parallel-comm` is enabled (single shared CUDA event causes NCCL order mismatch)

## Summary
Enabling `--overlap-moe-expert-parallel-comm` can deadlock at scale. The overlap scheduler currently uses **one shared CUDA event** as a “baton” for *both* compute and comm streams across all nodes in a model chunk. This lets **multiple comm ops become eligible at once** on each rank, so different ranks may launch collectives in **different orders** (e.g., `[C1, C2]` vs `[C2, C1]`). That violates NCCL’s “same collectives, same order per communicator” invariant and leads to a hang.

## Affected code (current behavior)
- `megatron/core/pipeline_parallel/utils.py`  
  `stream_acquire_context(stream, event)` uses **the same event** for both `wait` and `record`.
- `megatron/core/models/common/model_chunk_schedule_plan.py`  
  A single `torch.cuda.Event()` is constructed and passed to **all** nodes (compute and comm). Both comm nodes (`moe_dispatch`, `moe_combine`) run on the comm stream and **both** gate on the same event.

This makes it possible for the run loop to enqueue **two comm sequences** like:
```
comm_stream: wait(E) ; NCCL #1 ; record(E)
comm_stream: wait(E) ; NCCL #2 ; record(E)
```
Once compute records `E`, *both* waits are satisfied and the two NCCL ops can start in **rank-local enqueue order**, which can differ across ranks → deadlock.

## Root cause (why it deadlocks)
- **NCCL invariant:** all ranks must call the **same collectives in the same order** for a given communicator.
- With a **single shared event**:
  1. Compute records `E`.
  2. Multiple comm nodes (e.g., `dispatch(k)` and `combine(k-1)` or `dispatch(k)` and `dispatch(k+1)`) were already queued behind `wait(E)`.
  3. Tiny host/GPU timing differences cause **different enqueue orders** across ranks.
  4. Each rank launches a different “first” collective → **order mismatch** → hang.

## Proposed fix
Replace the single baton with **two CUDA events** per model chunk and make the contract explicit:

- **Two events per chunk**
  - `comp_event`: recorded by compute nodes.
  - `comm_event`: recorded by comm nodes.

- **Alternating dependency**
  - **Compute nodes:** `wait(comm_event)` → compute → `record(comp_event)`
  - **Comm nodes:** `wait(comp_event)` → NCCL all-to-all → `record(comm_event)`

- **Seed once**  
  On init, record `comm_event` on the compute stream once so the first compute step isn’t blocked.

- **APIs**
  - Change `stream_acquire_context` to `stream_acquire_context(stream, wait_event, record_event)`.
  - `ScheduleNode(..., event_pair=(wait_event, record_event))` (remove single-event legacy).

This preserves compute/comm overlap but prevents “two comm ops eligible at once” races.

**To Reproduce**
Steps to reproduce the behavior. The easier it is to reproduce the faster it will get maintainer attention.

Train a large MoE (like the DeepSeek 671B using the default configs from Megatron Model Zoo on a large number of GPUs (128 nodes per the configuration). It will reliably deadlock when using the latest main when pretraining from scratch as the computation between experts is more unbalanced near the beginning of training. I am trying to craft a small repro that occurs on a small number of GPUs, but I have been unsuccessful so far. The issue occurs on H100s and H200s.

**Expected behavior**

**Stack trace/logs**
If applicable, add the stack trace or logs from the time of the error.

**Environment (please complete the following information):**
 - Megatron-LM commit ID - 15a0d4738f062b93b0f698dfc0a3b1bb63f26878
 - PyTorch version - Nemo 25.07 container
 - CUDA version - 12.9
 - NCCL version - version associated with Nemo 25.07 container

**Proposed fix**
If you have a proposal for how to fix the issue state it here or link to a PR.

**Additional context**

**Comment by yanring** [comment]
Adding @Wohox @lhb8125

**Comment by Skylion007** [comment]
not sure if this is the root cause analysis is correct here, but I do think that the code is likely triggering some failure of the NCCL invariant at the beginning of the first backward pass with a lot of ranks

**Comment by lhb8125** [comment]
@Skylion007 I may not get your idea, could you describe the issue with a specific workflow?
like in the following workflow, which two comm kernels will run in random order?
```
comm_stream: combine_bwd            | dispatch_fwd->dispatch_bwd  | combine_fwd
comp_stream: attn_fwd->post_attn_fwd| mlp_bwd->mlp_bwd_dw->mlp_fwd| post_attn_bwd->attn_bwd
```

**Comment by Wohox** [comment]
Can you also share the nsys/torch timeline/screenshot that demonstrates the mismatched launch order if possible?

**Comment by Skylion007** [comment]
Okay, so more details about these deadlocks. I am actually not sure the dual baton issue I posted above is necessarily the issue after all, but there is some collective deadlock. @Wohox Any instructions on how to best extract the deadlock order out of this many GPUs?


I also did find out something interesting after trying different `--pipeline-model-parallel-layout`. It looks like any pipeline-model-parallel-layout that has an odd number of layers in one of the pipeline stages will deadlock with the scheduler.

```
A2A_OVERLAP=1 PP=8 VPP=4 TP=2 EP=64 NNODES=128 GBS=8192 PR=fp8 bash sbatch_benchmarking.sh \    --recompute-granularity selective  --recompute-modules mla_up_proj mlp  --pipeline-model-parallel-layout "Et|(tt|)*30mL"
```
Works! 

```
A2A_OVERLAP=1 PP=8 VPP=4 TP=2 EP=64 NNODES=128 GBS=8192 PR=fp8 bash sbatch_benchmarking.sh \    --recompute-granularity selective   --recompute-modules mla_up_proj mlp  --pipeline-model-parallel-layout "Et*2|(tt|)*22t|(tt|)*7mL" # (from DeepEP benchmarking GH issue)
```
Deadlocks!

```
A2A_OVERLAP=1 PP=8 VPP=4 TP=2 EP=64 NNODES=128 GBS=8192 PR=fp8 bash sbatch_benchmarking.sh \    --recompute-granularity selective \    --recompute-modules mla_up_proj mlp \    --pipeline-model-parallel-layout ""Et|(tt|)*29t|mL"" # from the earlier benchmarking script @Zhongbo Zhu attached
```
Deadlocks!

**Comment by Wohox** [comment]
@Skylion007 I failed to reproduce the hang issue by running the following command (multiple runs, using NEMO25.07, commit https://github.com/NVIDIA/Megatron-LM/commit/15a0d4738f062b93b0f698dfc0a3b1bb63f26878)
```bash
A2A_OVERLAP=1 PP=8 VPP=4 TP=2 EP=64 NNODES=128 GBS=8192 PR=fp8 bash sbatch_benchmarking.sh \    --recompute-granularity selective   --recompute-modules mla_up_proj mlp  --pipeline-model-parallel-layout "Et*2|(tt|)*22t|(tt|)*7mL" # (from DeepEP benchmarking GH issue)
```
My suspicion it that the flags are not exactly the same between my experiments and yours, for example, Nemo25.07 doesn't contain TE2.8, therefore `--overlap-grad-reduce --overlap-param-gather` should be disabled (enabled in modelZoo). Therefore I will attach the log here for comparison and please check for any mismatched flags so that we can reproduce the issue locally.

cc @lhb8125 
[3699342.log](https://github.com/user-attachments/files/22425488/3699342.log)

**Comment by LiJunscs** [comment]
@Skylion007 I encountered the same problem as you did. However, even after making modifications according to your suggestions, it still gets stuck. Could you provide more detailed modification advice?

**Comment by LiJunscs** [comment]
@Wohox Hi, I reproduce the bug almost evrey time, here is my train.sh. I run with a A800 cluster. My Megatron version is core_0.14.0 release and TE version is 2.6.0, Pytorch version is 2.4.0. And I tried the advice @Skylion007 metioned, but it still stuck.
```
#!/bin/bash

# Environment variables for performance tuning
export CUDA_DEVICE_MAX_CONNECTIONS=${CUDA_DEVICE_MAX_CONNECTIONS:-1}
#export LOG_LEVEL=${LOG_LEVEL:-INFO}
#export NCCL_IB_TIMEOUT=${NCCL_IB_TIMEOUT:-19}
#export NVTE_FWD_LAYERNORM_SM_MARGIN=${NVTE_FWD_LAYERNORM_SM_MARGIN:-16}
#export NVTE_BWD_LAYERNORM_SM_MARGIN=${NVTE_BWD_LAYERNORM_SM_MARGIN:-16}
#export NCCL_P2P_NET_CHUNKSIZE=${NCCL_P2P_NET_CHUNKSIZE:-2097152}
#export NCCL_AVOID_RECORD_STREAMS=${NCCL_AVOID_RECORD_STREAMS:-1}
# export NCCL_DEBUG=INFO

CHECKPOINT_PATH=${1:-"checkpoints/llama3_8b"}
TENSORBOARD_LOGS_PATH=${2:-"tensorboard_logs/llama3_8b"}
# TOKENIZER_ARG=${3:-"MOCK"} # Path to tokenizer model, or "MOCK"
# DATA_ARG=${4:-"MOCK"}     # Data prefix, or "MOCK"
TOKENIZER_ARG="/xxx/models/llama3"
DATA_ARG="/xxx/datasets/mcore/wiki_text_document"

# Create directories if they don't exist
mkdir -p "$(dirname "$CHECKPOINT_PATH")"
mkdir -p "$(dirname "$TENSORBOARD_LOGS_PATH")"

# Distributed training setup
GPUS_PER_NODE=8
NUM_NODES=1
MASTER_ADDR=${MASTER_ADDR:-localhost}
MASTER_PORT=${MASTER_PORT:-6000}
NODE_RANK=${NODE_RANK:-0}
WORLD_SIZE=$(($GPUS_PER_NODE*$NUM_NODES))

# Path to the pretrain_gpt.py script, assuming this script is run from the root of the Megatron-LM repository
PRETRAIN_SCRIPT_PATH="pretrain_gpt.py"

# Fixed model and training parameters
TP_SIZE=1     
CP_SIZE=1     
PP_SIZE=4
VP_SIZE=2     
MICRO_BATCH_SIZE=1
GLOBAL_BATCH_SIZE=16
NUM_LAYERS=16  
DTYPE="bf16"
SEQ_LENGTH=8192
MAX_POSITION_EMBEDDINGS=8192

# Data cache path (useful for both mock and real data)
DATA_CACHE_PATH="${PWD}/benchmark_cache_llama3_8b_fp8"
mkdir -p "$DATA_CACHE_PATH"

DISTRIBUTED_ARGS=(
    --nproc_per_node $GPUS_PER_NODE
    --nnodes $NUM_NODES
    --node_rank $NODE_RANK
    --master_addr $MASTER_ADDR
    --master_port $MASTER_PORT
)

MODEL_ARGS=(
    --use-mcore-models
    --num-layers $NUM_LAYERS
    --hidden-size 4096
    --ffn-hidden-size 14336
    --num-attention-heads 32
    --group-query-attention
    --num-query-groups 8
    --kv-channels 128
    --seq-length $SEQ_LENGTH
    --max-position-embeddings $MAX_POSITION_EMBEDDINGS
    --position-embedding-type rope
    --rotary-base 1000000 
    --rotary-percent 1.0
    --attention-dropout 0.0
    --hidden-dropout 0.0
    --swiglu
    --init-method-std 0.0134
    --attention-backend fused
    --apply-layernorm-1p 
    --untie-embeddings-and-output-weights
    --disable-bias-linear 
    --no-rope-fusion
    # --transformer-impl local
    --normalization RMSNorm
    --no-persist-layer-norm
    --num-experts 8
    --moe-router-topk 1
    --moe-ffn-hidden-size 4096
    --moe-grouped-gemm
)

TRAINING_ARGS=(
    --micro-batch-size $MICRO_BATCH_SIZE
    --global-batch-size $GLOBAL_BATCH_SIZE
    # --train-samples 1953125000
    # --lr-decay-samples 1949218748
    # --lr-warmup-samples 3906252
    --train-iters 10
    --lr 0.00015
    --min-lr 0.00001
    --decoupled-lr 5.0e-4      # Specific to decoupled AdamW, ensure optimizer is compatible
    --decoupled-min-lr 4.5e-5  # Specific to decoupled AdamW
    --lr-decay-style cosine
    --clip-grad 1.0
    --weight-decay 0.1
    --adam-beta1 0.9
    --adam-beta2 0.95
    --bf16
    --grad-reduce-in-bf16
    --cross-entropy-loss-fusion
    --calculate-per-token-loss 
    --manual-gc 
    --empty-unused-memory-level 1 
    --exit-duration-in-mins 235 
    --no-gradient-accumulation-fusion
    --no-masked-softmax-fusion 
    --attention-backend flash
    --overlap-grad-reduce
    --moe-token-dispatcher-type alltoall
    --moe-router-pre-softmax
    --overlap-moe-expert-parallel-comm
    # --moe-enable-deepep
    # --delay-wgrad-compute
    # --moe-router-dtype fp32
)

# Conditional arguments based on DTYPE (FP8)
DTYPE_ARGS=()
if [[ "$DTYPE" == "fp8" ]]; then
    DTYPE_ARGS+=(
        "--fp8-format hybrid"
        "--fp8-amax-history-len 1024"
        "--fp8-amax-compute-algo max"
        "--fp8-param-gather"
    )
fi

# Model parallelism arguments
MODEL_PARALLEL_ARGS=(
    --tensor-model-parallel-size $TP_SIZE
    --context-parallel-size $CP_SIZE
    --pipeline-model-parallel-size $PP_SIZE # Not explicitly set in llama script options, assume 1 if not multi-node PP
    --sequence-parallel  # Always enable sequence parallelism with TP_SIZE=2
    --num-virtual-stages-per-pipeline-rank $VP_SIZE
    --expert-model-parallel-size 2
)

# Distributed Data Parallel (DDP) arguments
# From original script's ddp_args
DDP_ARGS=(
    --use-distributed-optimizer
    --overlap-grad-reduce
    --overlap-param-gather
)
TRAINING_ARGS+=("${DDP_ARGS[@]}")


# Data arguments (conditional for mock vs real data)
DATA_ARGS_LIST=()
if [[ "$TOKENIZER_ARG" == "MOCK" ]] || [[ "$DATA_ARG" == "MOCK" ]] || [[ -z "$TOKENIZER_ARG" ]]; then
    DATA_ARGS_LIST+=(
        "--mock-data"
        "--tokenizer-type NullTokenizer"
        "--vocab-size 128256" 
        "--data-cache-path ${DATA_CACHE_PATH}"
        "--tiktoken-pattern v2" 
        "--split '99,1,0'"
        "--no-create-attention-mask-in-dataloader"
        "--no-mmap-bin-files"
        "--num-workers 1"
    )
else
    # Settings for real data
    DATA_ARGS_LIST+=(
        "--data-path $DATA_ARG"
        "--tokenizer-type HuggingFaceTokenizer" 
        "--tokenizer-model $TOKENIZER_ARG"
        "--data-cache-path ${DATA_CACHE_PATH}"
        "--split '99,1,0'"
        "--no-create-attention-mask-in-dataloader"
        "--no-mmap-bin-files"
        "--num-workers 1"
        # Note: --vocab-size might be inferred by HuggingFaceTokenizer or might need to be explicit.
        "--vocab-size 128256"
    )
fi

EVAL_AND_LOGGING_ARGS=(
    --log-interval 1
    --eval-iters 32
    --eval-interval 100
    --save-interval 1000
    --log-throughput
    # --profile
    # --profile-step-start 4
    # --profile-step-end 6
    # --use-torch-profiler
    --ckpt-format torch_dist 
    --distributed-timeout-minutes 60
    --save "$CHECKPOINT_PATH"
    --load "$CHECKPOINT_PATH" 
    --tensorboard-dir "$TENSORBOARD_LOGS_PATH"
)


# METHOD="1f1b"
METHOD="interleaved_1f1b"
LOGGING_ARGS=(
    --eval-iters 0
    --log-interval 1
    --log-throughput
    --profile-step-start 0
    --profile-step-end 9
    --profile-ranks 0 1 2 3 4 5 6 7
    --tensorboard-dir "${TENSORBOARD_LOGS_PATH}/${METHOD}"
)

# Ensure pretrain_gpt.py is found
if [ ! -f "$PRETRAIN_SCRIPT_PATH" ]; then
    echo "Error: pretrain_gpt.py not found at $PRETRAIN_SCRIPT_PATH"
    echo "Please ensure you are running this script from the root of the Megatron-LM repository, and pretrain_gpt.py is present."
    exit 1
fi

mkdir -p "${TENSORBOARD_LOGS_PATH}/${METHOD}"
rm -f "${TENSORBOARD_LOGS_PATH}/${METHOD}/*"
TIMESTAMP=$(date +"%Y-%m-%d_%H-%M-%S")
LOG_FILE="logs/${TIMESTAMP}_${METHOD}_PP${PP_SIZE}_TP${TP_SIZE}_DP${DP_SIZE}_CP${CP_SIZE}_EP${EP_SIZE}_VP${VP_SIZE}.log"


# nsys command
NSYS_ITER=-1
if (( $NSYS_ITER >= 0 )); then
    mkdir -p "./logs/nsys_reports/${METHOD}"
    NSYS_CMD="nsys profile -s none -t nvtx,cuda --force-overwrite true -o ./logs/nsys_reports/${METHOD}/${NODE_RANK} --capture-range=cudaProfilerApi --capture-range-end=stop"
else
    NSYS_CMD=""
fi

export ENABLE_DEBUG=false
# Run the training command
${NSYS_CMD} torchrun ${DISTRIBUTED_ARGS[@]} \
    "$PRETRAIN_SCRIPT_PATH" \
    ${MODEL_ARGS[@]} \
    ${TRAINING_ARGS[@]} \
    ${DTYPE_ARGS[@]} \
    ${MODEL_PARALLEL_ARGS[@]} \
    ${DATA_ARGS_LIST[@]} \
    ${LOGGING_ARGS[@]} \
    | tee "$LOG_FILE" 2>&1
    # ${EVAL_AND_LOGGING_ARGS[@]}

set +x
```
